glommio/src/channels/channel_mesh.rs

```diff
@@ -0,0 +1,312 @@
+            let receiver_provider = ReceiverProvider::Shared(receiver);
+            (Cell::new(sender_provider), Cell::new(receiver_provider))
+        }
+    }
+
+    async fn join(self) -> (Senders<T>, Receivers<T>) {
+        let mut notifiers = self.notifiers.write().unwrap();
+
+        if notifiers.len() == self.nr_peers {
+            panic!("The channel mesh is full.");
+        }
+
+        let index = notifiers
+            .binary_search_by(|n| n.executor_id.cmp(&Local::id()))
+            .expect_err("Should not join a mesh more than once.");
+
+        let peer_id = if notifiers.len() == self.nr_peers - 1 {
+            notifiers.insert(index, Notifier::new(None));
+            for (peer_id, notifier) in notifiers.iter_mut().enumerate() {
+                if let Some(sender) = notifier.sender.take() {
+                    sender.connect().await.send(peer_id).await.unwrap();
+                }
+            }
+            drop(notifiers);
+            index
+        } else {
+            let (sender, receiver) = shared_channel::new_bounded(1);
+            notifiers.insert(index, Notifier::new(Some(sender)));
+            drop(notifiers);
+            receiver.connect().await.recv().await.unwrap()
+        };
+
+        let providers = self.providers.as_ref();
+
+        let mut senders = Senders {
+            id: peer_id,
+            senders: Vec::with_capacity(self.nr_peers),
+        };
+        let mut receivers = Receivers {
+            id: peer_id,
+            receivers: Vec::with_capacity(self.nr_peers),
+        };
+
+        for i in 0..self.nr_peers {
+            let sender = providers[peer_id][i].0.take();
+            let receiver = providers[i][peer_id].1.take();
+            let (sender, receiver) = future::join(sender.get(), receiver.get()).await;
+            senders.senders.push(sender);
+            receivers.receivers.push(receiver);
+        }
+
+        (senders, receivers)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::enclose;
+    use crate::prelude::*;
+
+    use super::*;
+
+    #[test]
+    fn test_channel_mesh() {
+        do_test_channel_mesh(0, 0, 0);
+        do_test_channel_mesh(1, 1, 1);
+        do_test_channel_mesh(10, 10, 10);
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_join_more_than_once() {
+        let mesh_builder = MeshBuilder::<u32>::new(2, 1);
+
+        LocalExecutor::default().run(enclose!((mesh_builder) async move {
+            future::join(mesh_builder.clone().join(), mesh_builder.join()).await;
+        }));
+    }
+
+    #[test]
+    #[should_panic]
+    fn test_join_more_than_capacity() {
+        do_test_channel_mesh(1, 1, 2);
+    }
+
+    fn do_test_channel_mesh(nr_peers: usize, channel_size: usize, nr_executors: usize) {
+        let mesh_builder = MeshBuilder::new(nr_peers, channel_size);
+
+        let executors = (0..nr_executors).map(|_| {
+            LocalExecutorBuilder::new().spawn(enclose!((mesh_builder) move || async move {
```

Comment:
What I think I am missing is the relationship between the mesh peer id with the executor id.

How am I supposed to use this in a real life scenario? Let's say that I am implementing a sharded key-value store.
Usually I will hash the key with some hashing algorithm, then % by the number of shards, then send to "shard 2" if `x % y == 2`.

But I see no translation happening between the mesh peer id and any id that the user could refer to.

Reply:
Btw, a very simple memcached-like sharded key-value is a great thing to add to examples if you are up to it!

Reply:
> What I think I am missing is the relationship between the mesh peer id with the executor id.
> 
> How am I supposed to use this in a real life scenario? Let's say that I am implementing a sharded key-value store.
> Usually I will hash the key with some hashing algorithm, then % by the number of shards, then send to "shard 2" if `x % y == 2`.
> 
> But I see no translation happening between the mesh peer id and any id that the user could refer to.

Peer ids are intended to be used for sharding directly. Translation is not necessary. I'll add some documentation to make it clear.

Reply:
> Btw, a very simple memcached-like sharded key-value is a great thing to add to examples if you are up to it!

There's a problem when I'm trying to write such an example. Messages passed through shared channels needs to be `Copy `, which is not possible for things like `String`. A walkaround  would be to wrap it with `Arc`. But I'm afraid it would introduce another memory indirection without benefit, which I'm not sure whether the compiler will optimize away. Would it be better to support non-copyable messages with a little bit of unsafe code (`ptr::read`, like the way in `VecDeque`) when trying to pop messages from the spsc_queue buffer?

Reply:
Because it is just an example we can use integers.
Passing strings is the wrong thing to do anyway because you will incur a copy which can be expensive (imagine you have a 1MB blob as a value, you don't want to be copying that around in the channel).

What you do in this situation is pass a pointer, and have some facility to make sure that the original data structure is alive.

My implementation of the 1:M ( I said N:M, but my use case really is more like 1:M) is a request-reply channel. I can send requests, and receive replies.

First, you should consider adding that to your implementation, if not always, at least optionally: If we want to be generic, in most times where we pass data to the other executor we will like to know if the operation on the data was successful. (Imagine you are passing buffers for another executor to do I/O on: you want to know if the I/O succeeded).

But now because the channel is request-reply, I keep the  big object (String) until the reply comes.
So you can pass a pointer through the channel, and the receiver knows that it will be alive until the reply is sent.

Reply:
An even more obvious example of why request-reply is beneficial:
- you send a `GET` request to one executor
- now you need to reply =)

Reply:
The N:M model is now supported as partial mesh, where a role of either producer or consumer can be specified when joining the mesh.

Reply:
> Passing strings is the wrong thing to do anyway because you will incur a copy which can be expensive (imagine you have a 1MB blob as a value, you don't want to be copying that around in the channel).

Rust `String` just wraps `Vec<u8>`, which is then implemented as a memory block pointed by a `Unique`, with an additional field for length. By using `ptr::write` and `ptr::read`, we can move a shallow copy of a string to the receiving thread without copying the contents, and lift the weight of managing raw pointers in client code. And since the ownership is also passed to the receiving thread, resources will be released there after operation, so no reply will be necessary in this scenario.